// piral-cli/lib/apps/validate-piral.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validatePiral = exports.validatePiralDefaults = void 0;
const common_1 = require("../common");
const common_2 = require("../common");
const rules_1 = require("../rules");
const types_1 = require("../types");
exports.validatePiralDefaults = {
    entry: './',
    logLevel: types_1.LogLevels.info,
};
function validatePiral(baseDir = process.cwd(), options = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        const { entry = exports.validatePiralDefaults.entry, logLevel = exports.validatePiralDefaults.logLevel } = options;
        common_2.setLogLevel(logLevel);
        common_2.progress('Reading configuration ...');
        const rules = yield rules_1.getPiralRules();
        const entryFiles = yield common_1.retrievePiralRoot(baseDir, entry);
        const _a = yield common_1.retrievePiletsInfo(entryFiles), { root, dependencies, ignored: _ } = _a, info = __rest(_a, ["root", "dependencies", "ignored"]);
        const errors = [];
        const warnings = [];
        yield common_2.checkCliCompatibility(root);
        const context = {
            error(message) {
                errors.push(common_2.log('generalError_0002', message));
            },
            warning(message) {
                warnings.push(common_2.log('generalWarning_0001', message));
            },
            logLevel,
            entry: entryFiles,
            dependencies: dependencies.std,
            devDependencies: dependencies.dev,
            peerDependencies: dependencies.peer,
            root,
            info,
        };
        yield common_1.runRules(rules, context, common_1.config.validators);
        common_1.ruleSummary(errors, warnings);
    });
}
exports.validatePiral = validatePiral;
