// piral-core/esm/utils/helpers.js
import { __assign, __rest, __spreadArray } from "tslib";
// tslint:disable-next-line
export var removeIndicator = null;
export function prependItem(items, item) {
    return __spreadArray([item], (items || []));
}
export function appendItem(items, item) {
    return __spreadArray(__spreadArray([], (items || [])), [item]);
}
export function prependItems(items, newItems) {
    return __spreadArray(__spreadArray([], newItems), (items || []));
}
export function appendItems(items, newItems) {
    return __spreadArray(__spreadArray([], (items || [])), newItems);
}
export function excludeItem(items, item) {
    return (items || []).filter(function (m) { return m !== item; });
}
export function includeItem(items, item) {
    return appendItem(excludeItem(items, item), item);
}
export function replaceOrAddItem(items, item, predicate) {
    var newItems = __spreadArray([], (items || []));
    for (var i = 0; i < newItems.length; i++) {
        if (predicate(newItems[i])) {
            newItems[i] = item;
            return newItems;
        }
    }
    newItems.push(item);
    return newItems;
}
export function removeNested(obj, predicate) {
    return Object.keys(obj).reduce(function (entries, key) {
        var item = obj[key];
        entries[key] = Object.keys(item).reduce(function (all, key) {
            var value = item[key];
            if (Array.isArray(value)) {
                all[key] = excludeOn(value, predicate);
            }
            else if (!value || !predicate(value)) {
                all[key] = value;
            }
            return all;
        }, {});
        return entries;
    }, {});
}
export function excludeOn(items, predicate) {
    return (items || []).filter(function (m) { return !predicate(m); });
}
export function updateKey(obj, key, value) {
    return value === removeIndicator ? withoutKey(obj, key) : withKey(obj, key, value);
}
export function withKey(obj, key, value) {
    var _a;
    return __assign(__assign({}, obj), (_a = {}, _a[key] = value, _a));
}
export function withoutKey(obj, key) {
    var _a = obj || {}, _b = key, _ = _a[_b], newObj = __rest(_a, [typeof _b === "symbol" ? _b : _b + ""]);
    return newObj;
}
