// piral-core/lib/state/createGlobalState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGlobalState = void 0;
var react_atom_1 = require("@dbeining/react-atom");
var react_router_dom_1 = require("react-router-dom");
var components_1 = require("../components");
function extend(defaultState, customState) {
    for (var _i = 0, _a = Object.keys(customState); _i < _a.length; _i++) {
        var key = _a[_i];
        if (key === '__proto__' || key === 'constructor') {
            continue;
        }
        var value = customState[key];
        var original = defaultState[key];
        var nested = typeof original === 'object' && typeof value === 'object';
        defaultState[key] = nested ? extend(original, value) : value;
    }
    return defaultState;
}
function createGlobalState(customState) {
    if (customState === void 0) { customState = {}; }
    var defaultState = {
        app: {
            error: undefined,
            loading: typeof window !== 'undefined',
            layout: 'desktop',
        },
        components: {
            ErrorInfo: components_1.DefaultErrorInfo,
            LoadingIndicator: components_1.DefaultLoadingIndicator,
            Router: react_router_dom_1.BrowserRouter,
            Layout: components_1.DefaultLayout,
        },
        errorComponents: {},
        registry: {
            extensions: {},
            pages: {},
            wrappers: {},
        },
        routes: {},
        data: {},
        portals: {},
        modules: [],
    };
    return react_atom_1.Atom.of(extend(defaultState, customState));
}
exports.createGlobalState = createGlobalState;
