// piral-dashboard/lib/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDashboardApi = void 0;
var tslib_1 = require("tslib");
var actions = require("./actions");
var piral_core_1 = require("piral-core");
var default_1 = require("./default");
function getPreferences(defaultPreferences, customPreferences) {
    if (customPreferences === void 0) { customPreferences = {}; }
    return tslib_1.__assign(tslib_1.__assign({}, defaultPreferences), customPreferences);
}
function getTiles(items, defaultPreferences) {
    var tiles = {};
    var i = 0;
    for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
        var _a = items_1[_i], component = _a.component, preferences = _a.preferences;
        tiles["global-" + i++] = {
            pilet: undefined,
            component: component,
            preferences: getPreferences(defaultPreferences, preferences),
        };
    }
    return tiles;
}
/**
 * Creates the Pilet API extension for activating dashboard support.
 */
function createDashboardApi(config) {
    if (config === void 0) { config = {}; }
    var _a = config.tiles, tiles = _a === void 0 ? [] : _a, _b = config.defaultPreferences, defaultPreferences = _b === void 0 ? {} : _b;
    return function (context) {
        context.defineActions(actions);
        context.dispatch(function (state) { return (tslib_1.__assign(tslib_1.__assign({}, state), { components: tslib_1.__assign({ DashboardTile: default_1.DefaultTile, DashboardContainer: default_1.DefaultContainer }, state.components), registry: tslib_1.__assign(tslib_1.__assign({}, state.registry), { tiles: getTiles(tiles, defaultPreferences) }) })); });
        return function (api, target) {
            var pilet = target.name;
            var next = 0;
            return {
                registerTile: function (name, arg, preferences) {
                    if (typeof name !== 'string') {
                        preferences = arg;
                        arg = name;
                        name = next++;
                    }
                    var id = piral_core_1.buildName(pilet, name);
                    context.registerTile(id, {
                        pilet: pilet,
                        component: piral_core_1.withApi(context, arg, api, 'tile'),
                        preferences: getPreferences(defaultPreferences, preferences),
                    });
                    return function () { return api.unregisterTile(name); };
                },
                unregisterTile: function (name) {
                    var id = piral_core_1.buildName(pilet, name);
                    context.unregisterTile(id);
                },
            };
        };
    };
}
exports.createDashboardApi = createDashboardApi;
