// piral-cli/lib/common/clients/npm.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.listPackage = exports.findTarball = exports.findSpecificVersion = exports.publishPackage = exports.createPackage = exports.installPackage = exports.unpackPackage = exports.installDependencies = void 0;
const path_1 = require("path");
const log_1 = require("../log");
const scripts_1 = require("../scripts");
const MemoryStream_1 = require("../MemoryStream");
function runNpmProcess(args, target, output) {
    log_1.log('generalDebug_0003', 'Starting the NPM process ...');
    const cwd = path_1.resolve(process.cwd(), target);
    return scripts_1.runCommand('npm', args, cwd, output);
}
function installDependencies(target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['install', '--legacy-peer-deps', ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM install dependencies result: ${ms.value}`);
        return ms.value;
    });
}
exports.installDependencies = installDependencies;
function unpackPackage(packageRef, target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['pack', packageRef, ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM (un)pack result: ${ms.value}`);
        return ms.value;
    });
}
exports.unpackPackage = unpackPackage;
function installPackage(packageRef, target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['install', packageRef, '--legacy-peer-deps', ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM install package result: ${ms.value}`);
        return ms.value;
    });
}
exports.installPackage = installPackage;
function createPackage(target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['pack', ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM pack result: ${ms.value}`);
        return ms.value;
    });
}
exports.createPackage = createPackage;
function publishPackage(target = '.', file = '*.tgz', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['publish', file, ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM publish result: ${ms.value}`);
        return ms.value;
    });
}
exports.publishPackage = publishPackage;
function findSpecificVersion(packageName, version) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['show', packageName, 'version', '--tag', version], '.', ms);
        log_1.log('generalDebug_0003', `NPM show result: ${ms.value}`);
        return ms.value;
    });
}
exports.findSpecificVersion = findSpecificVersion;
function findTarball(packageRef, target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        yield runNpmProcess(['view', packageRef, 'dist.tarball', ...flags], target, ms);
        log_1.log('generalDebug_0003', `NPM view packageRef result: ${ms.value}`);
        return ms.value;
    });
}
exports.findTarball = findTarball;
function listPackage(packageRef, target = '.', ...flags) {
    return __awaiter(this, void 0, void 0, function* () {
        const ms = new MemoryStream_1.MemoryStream();
        try {
            yield runNpmProcess(['ls', packageRef, '--json', '--depth', '0', ...flags], target, ms);
        }
        catch (e) {
            log_1.log('generalDebug_0003', `NPM ls packageRef error: ${e}`);
        }
        log_1.log('generalDebug_0003', `NPM ls packageRef result: ${ms.value}`);
        return JSON.parse(ms.value);
    });
}
exports.listPackage = listPackage;
