// piral-core/lib/createInstance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstance = void 0;
var tslib_1 = require("tslib");
var piral_base_1 = require("piral-base");
var modules_1 = require("./modules");
var state_1 = require("./state");
var helpers_1 = require("./helpers");
/**
 * Creates a new PiralInstance component, which can be used for
 * bootstrapping the application easily.
 *
 * @example
```jsx
const instance = createInstance({
  requestPilets() {
    return fetch(...);
  },
});

const app = (
  <Piral instance={instance}>
    <Define name="Layout" component={MyLayout} />
  </Piral>
);
render(app, document.querySelector('#app'));
```
 */
function createInstance(config) {
    if (config === void 0) { config = {}; }
    var state = config.state, actions = config.actions, _a = config.availablePilets, availablePilets = _a === void 0 ? [] : _a, extendApi = config.extendApi, plugins = config.plugins, _b = config.requestPilets, requestPilets = _b === void 0 ? modules_1.defaultModuleRequester : _b, fetchDependency = config.fetchDependency, _c = config.getDependencies, getDependencies = _c === void 0 ? modules_1.getLocalDependencies : _c, loaderConfig = config.loaderConfig, _d = config.async, async = _d === void 0 ? false : _d, loadPilet = config.loadPilet, loaders = config.loaders, _e = config.apiFactory, apiFactory = _e === void 0 ? modules_1.defaultApiFactory : _e;
    var globalState = state_1.createGlobalState(state);
    var events = piral_base_1.createListener(globalState);
    var context = state_1.createActions(globalState, events);
    var definedPlugins = plugins || extendApi || [];
    var usedPlugins = Array.isArray(definedPlugins) ? definedPlugins : [definedPlugins];
    var createApi = apiFactory(context, usedPlugins);
    var root = createApi({
        name: 'root',
        version: process.env.BUILD_PCKG_VERSION || '1.0.0',
        spec: '',
    });
    var options = helpers_1.createPiletOptions({
        context: context,
        createApi: createApi,
        loaders: loaders,
        loadPilet: loadPilet,
        availablePilets: availablePilets,
        fetchDependency: fetchDependency,
        loaderConfig: loaderConfig,
        getDependencies: getDependencies,
        strategy: piral_base_1.isfunc(async) ? async : async ? piral_base_1.blazingStrategy : piral_base_1.standardStrategy,
        requestPilets: requestPilets,
    });
    if (actions) {
        state_1.includeActions(context, actions);
    }
    context.options = options;
    return tslib_1.__assign(events, {
        createApi: createApi,
        context: context,
        root: root,
        options: options,
    });
}
exports.createInstance = createInstance;
