// piral-base/esm/dependency.js
function requireModule(name, dependencies) {
    var dependency = dependencies[name];
    if (!dependency) {
        var error = new Error("Cannot find module '" + name + "'");
        error.code = 'MODULE_NOT_FOUND';
        throw error;
    }
    return dependency;
}
function checkPiletApp(name, app) {
    if (!app) {
        console.error('Invalid module found.', name);
    }
    else if (typeof app.setup !== 'function') {
        console.warn('Setup function is missing.', name);
    }
    else {
        return app;
    }
    return {
        setup: function () { },
    };
}
function checkPiletAppAsync(name, app) {
    return Promise.resolve(app).then(function (resolvedApp) { return checkPiletApp(name, resolvedApp); });
}
function getLocalRequire(dependencies) {
    if (dependencies === void 0) { dependencies = {}; }
    return function (moduleName) { return requireModule(moduleName, dependencies); };
}
/**
 * Compiles the given content from a generic dependency.
 * @param name The name of the dependency to compile.
 * @param content The content of the dependency to compile.
 * @param link The optional link to the dependency.
 * @param dependencies The globally available dependencies.
 * @returns The evaluated dependency.
 */
export function evalDependency(name, content, link, dependencies) {
    if (link === void 0) { link = ''; }
    var mod = {
        exports: {},
    };
    var require = getLocalRequire(dependencies);
    try {
        var sourceUrl = link && "\n//# sourceURL=" + link;
        var importer = new Function('module', 'exports', 'require', content + sourceUrl);
        importer(mod, mod.exports, require);
    }
    catch (e) {
        console.error("Error while evaluating " + name + ".", e);
    }
    return mod.exports;
}
/**
 * Compiles the given content from a module with a dependency resolution.
 * @param name The name of the dependency to compile.
 * @param content The content of the dependency to compile.
 * @param link The optional link to the dependency.
 * @param dependencies The globally available dependencies.
 * @returns The evaluated module.
 */
export function compileDependency(name, content, link, dependencies) {
    if (link === void 0) { link = ''; }
    var app = evalDependency(name, content, link, dependencies);
    return checkPiletAppAsync(name, app);
}
function includeScript(piletName, depName, link, integrity, dependencies, crossOrigin) {
    window[depName] = getLocalRequire(dependencies);
    return includeScriptDependency(link, integrity, crossOrigin).then(function (s) { return checkPiletAppAsync(piletName, s.app); }, function () { return checkPiletApp(piletName); });
}
/**
 * Includes a dependency as a script.
 * @param link The link to the script.
 * @param integrity The integrity for the script, if any.
 * @param crossOrigin Defines if cross-origin should be used.
 * @returns The script element.
 */
export function includeScriptDependency(link, integrity, crossOrigin) {
    return new Promise(function (resolve, reject) {
        var s = document.createElement('script');
        s.async = true;
        s.src = link;
        if (integrity) {
            s.crossOrigin = crossOrigin || 'anonymous';
            s.integrity = integrity;
        }
        else if (crossOrigin) {
            s.crossOrigin = crossOrigin;
        }
        s.onload = function () { return resolve(s); };
        s.onerror = function () { return reject(); };
        document.body.appendChild(s);
    });
}
/**
 * Includes the given single pilet script via its URL with a dependency resolution.
 * @param meta The meta data of the dependency to include.
 * @param dependencies The globally available dependencies.
 * @param crossOrigin The override for the cross-origin attribute.
 * @returns The evaluated module.
 */
export function includeDependency(meta, dependencies, crossOrigin) {
    return includeScript(meta.name, meta.requireRef, meta.link, meta.integrity, dependencies, crossOrigin);
}
/**
 * Includes the given bundle script via its URL with a dependency resolution.
 * @param meta The meta data of the dependency to include.
 * @param dependencies The globally available dependencies.
 * @param crossOrigin The override for the cross-origin attribute.
 * @returns The evaluated module.
 */
export function includeBundle(meta, dependencies, crossOrigin) {
    var _a;
    return includeScript((_a = meta.name) !== null && _a !== void 0 ? _a : '(bundle)', meta.bundle, meta.link, meta.integrity, dependencies, crossOrigin);
}
