// piral/esm/run.js
import { __rest } from "tslib";
import { getAppInstance } from './app';
function noChange(config) {
    return config;
}
/**
 * Sets up a new Piral instance and runs it using the provided runner.
 * @param options The options to use when setting up the Piral instance.
 * @example
```ts
import { render } from 'react-dom';
import { runInstance } from 'piral';

runInstance((app, selector) => render(app, document.querySelector(selector)));
```
 */
export function runInstance(runner, options) {
    if (options === void 0) { options = {}; }
    var _a = options.selector, selector = _a === void 0 ? '#app' : _a, settings = options.settings, layout = options.layout, piralChildren = options.piralChildren, dashboardPath = options.dashboardPath, errors = options.errors, _b = options.middleware, middleware = _b === void 0 ? noChange : _b, config = __rest(options, ["selector", "settings", "layout", "piralChildren", "dashboardPath", "errors", "middleware"]);
    var _c = getAppInstance(middleware(config), {
        settings: settings,
        layout: layout,
        errors: errors,
        dashboardPath: dashboardPath,
        piralChildren: piralChildren,
    }), app = _c.app, instance = _c.instance;
    runner(app, selector);
    return instance;
}
