// piral/lib/run.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runInstance = void 0;
var tslib_1 = require("tslib");
var app_1 = require("./app");
function noChange(config) {
    return config;
}
/**
 * Sets up a new Piral instance and runs it using the provided runner.
 * @param options The options to use when setting up the Piral instance.
 * @example
```ts
import { render } from 'react-dom';
import { runInstance } from 'piral';

runInstance((app, selector) => render(app, document.querySelector(selector)));
```
 */
function runInstance(runner, options) {
    if (options === void 0) { options = {}; }
    var _a = options.selector, selector = _a === void 0 ? '#app' : _a, settings = options.settings, layout = options.layout, piralChildren = options.piralChildren, dashboardPath = options.dashboardPath, errors = options.errors, _b = options.middleware, middleware = _b === void 0 ? noChange : _b, config = tslib_1.__rest(options, ["selector", "settings", "layout", "piralChildren", "dashboardPath", "errors", "middleware"]);
    var _c = app_1.getAppInstance(middleware(config), {
        settings: settings,
        layout: layout,
        errors: errors,
        dashboardPath: dashboardPath,
        piralChildren: piralChildren,
    }), app = _c.app, instance = _c.instance;
    runner(app, selector);
    return instance;
}
exports.runInstance = runInstance;
