// piral-menu/esm/create.js
import { __assign } from "tslib";
import * as actions from './actions';
import { withApi, buildName } from 'piral-core';
import { DefaultContainer, DefaultItem } from './default';
function getSettings(defaultSettings, customSettings) {
    if (customSettings === void 0) { customSettings = {}; }
    return __assign(__assign({ type: 'general' }, defaultSettings), customSettings);
}
function getMenuItems(items, defaultSettings) {
    var menuItems = {};
    var i = 0;
    for (var _i = 0, items_1 = items; _i < items_1.length; _i++) {
        var _a = items_1[_i], component = _a.component, settings = _a.settings;
        menuItems["global-" + i++] = {
            pilet: undefined,
            component: component,
            settings: getSettings(defaultSettings, settings),
        };
    }
    return menuItems;
}
/**
 * Creates new Pilet API extensions for integration of menu items.
 */
export function createMenuApi(config) {
    if (config === void 0) { config = {}; }
    var _a = config.items, items = _a === void 0 ? [] : _a, _b = config.defaultSettings, defaultSettings = _b === void 0 ? {} : _b;
    return function (context) {
        context.defineActions(actions);
        context.dispatch(function (state) { return (__assign(__assign({}, state), { components: __assign({ MenuContainer: DefaultContainer, MenuItem: DefaultItem }, state.components), registry: __assign(__assign({}, state.registry), { menuItems: getMenuItems(items, defaultSettings) }) })); });
        return function (api, target) {
            var pilet = target.name;
            var next = 0;
            return {
                registerMenu: function (name, arg, settings) {
                    if (typeof name !== 'string') {
                        settings = arg;
                        arg = name;
                        name = next++;
                    }
                    var id = buildName(pilet, name);
                    context.registerMenuItem(id, {
                        pilet: pilet,
                        component: withApi(context, arg, api, 'menu'),
                        settings: getSettings(defaultSettings, settings),
                    });
                    return function () { return api.unregisterMenu(name); };
                },
                unregisterMenu: function (name) {
                    var id = buildName(pilet, name);
                    context.unregisterMenuItem(id);
                },
            };
        };
    };
}
