// piral-core/esm/state/createGlobalState.js
import { Atom } from '@dbeining/react-atom';
import { BrowserRouter } from 'react-router-dom';
import { DefaultErrorInfo, DefaultLoadingIndicator, DefaultLayout } from '../components';
function extend(defaultState, customState) {
    for (var _i = 0, _a = Object.keys(customState); _i < _a.length; _i++) {
        var key = _a[_i];
        if (key === '__proto__' || key === 'constructor') {
            continue;
        }
        var value = customState[key];
        var original = defaultState[key];
        var nested = typeof original === 'object' && typeof value === 'object';
        defaultState[key] = nested ? extend(original, value) : value;
    }
    return defaultState;
}
export function createGlobalState(customState) {
    if (customState === void 0) { customState = {}; }
    var defaultState = {
        app: {
            error: undefined,
            loading: typeof window !== 'undefined',
            layout: 'desktop',
        },
        components: {
            ErrorInfo: DefaultErrorInfo,
            LoadingIndicator: DefaultLoadingIndicator,
            Router: BrowserRouter,
            Layout: DefaultLayout,
        },
        errorComponents: {},
        registry: {
            extensions: {},
            pages: {},
            wrappers: {},
        },
        routes: {},
        data: {},
        portals: {},
        modules: [],
    };
    return Atom.of(extend(defaultState, customState));
}
