// piral-cli-webpack/lib/actions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildPilet = exports.debugPilet = exports.buildPiral = exports.watchPiral = exports.debugPiral = void 0;
const webpack_1 = require("./webpack");
exports.debugPiral = {
    run(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = yield webpack_1.callDynamic('debug-piral', args);
            return bundler;
        });
    },
};
exports.watchPiral = {
    run(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = yield webpack_1.callStatic('debug-mono-piral', args);
            return bundler;
        });
    },
};
exports.buildPiral = {
    run(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = yield webpack_1.callStatic('build-piral', args);
            return bundler.bundle;
        });
    },
};
exports.debugPilet = {
    run(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = yield webpack_1.callDynamic('debug-pilet', args);
            return bundler;
        });
    },
};
exports.buildPilet = {
    run(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const bundler = yield webpack_1.callStatic('build-pilet', args);
            return bundler.bundle;
        });
    },
};
