// piral-core/lib/components/components.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PiralDebug = exports.PiralLayout = exports.PiralRouter = exports.PiralLoadingIndicator = exports.PiralError = exports.getPiralComponent = void 0;
var tslib_1 = require("tslib");
var React = require("react");
var hooks_1 = require("../hooks");
function getPiralComponent(name) {
    return function (props) {
        var Component = hooks_1.useGlobalState(function (s) { return s.components[name]; });
        return Component ? React.createElement(Component, tslib_1.__assign({}, props)) : null;
    };
}
exports.getPiralComponent = getPiralComponent;
/**
 * Gets the currently registered ErrorInfo component.
 * By default the DefaultErrorInfo component is used.
 */
exports.PiralError = getPiralComponent('ErrorInfo');
/**
 * Gets the currently registered LoadingIndicator component.
 * By default only Loading is rendered.
 */
exports.PiralLoadingIndicator = getPiralComponent('LoadingIndicator');
/**
 * Gets the currently registered Router component.
 * By default the BrowserRouter is used.
 */
exports.PiralRouter = getPiralComponent('Router');
/**
 * Gets the currently registered Layout component.
 * By default the children are rendered.
 */
exports.PiralLayout = getPiralComponent('Layout');
/**
 * Gets the currently registered Debug component.
 * By default nothing is used.
 */
exports.PiralDebug = getPiralComponent('Debug');
