// piral-modals/esm/create.js
import { __assign } from "tslib";
import * as actions from './actions';
import { withApi, buildName } from 'piral-core';
import { DefaultHost, DefaultDialog } from './default';
function getModalDialogs(dialogs) {
    var modals = {};
    for (var _i = 0, dialogs_1 = dialogs; _i < dialogs_1.length; _i++) {
        var _a = dialogs_1[_i], name_1 = _a.name, component = _a.component, defaults = _a.defaults;
        modals["global-" + name_1] = {
            pilet: undefined,
            name: name_1,
            component: component,
            defaults: defaults,
        };
    }
    return modals;
}
/**
 * Creates new Pilet API extensions for support modal dialogs.
 */
export function createModalsApi(config) {
    if (config === void 0) { config = {}; }
    var _a = config.dialogs, dialogs = _a === void 0 ? [] : _a, _b = config.selectId, selectId = _b === void 0 ? function (name) { return name + "-" + ~~(Math.random() * 10000); } : _b;
    return function (context) {
        context.defineActions(actions);
        context.dispatch(function (state) { return (__assign(__assign({}, state), { components: __assign({ ModalsHost: DefaultHost, ModalsDialog: DefaultDialog }, state.components), registry: __assign(__assign({}, state.registry), { modals: getModalDialogs(dialogs) }), modals: [] })); });
        return function (api, target) {
            var pilet = target.name;
            return {
                showModal: function (simpleName, options) {
                    var name = buildName(pilet, simpleName);
                    var dialog = {
                        id: selectId(name),
                        name: name,
                        alternative: simpleName,
                        options: options,
                        close: function () {
                            setTimeout(function () { return context.closeModal(dialog); }, 0);
                        },
                    };
                    context.openModal(dialog);
                    return dialog.close;
                },
                registerModal: function (name, arg, defaults) {
                    var id = buildName(pilet, name);
                    context.registerModal(id, {
                        pilet: pilet,
                        name: name,
                        component: withApi(context, arg, api, 'modal'),
                        defaults: defaults,
                    });
                    return function () { return api.unregisterModal(name); };
                },
                unregisterModal: function (name) {
                    var id = buildName(pilet, name);
                    context.unregisterModal(id);
                },
            };
        };
    };
}
