// piral-core/esm/modules/api.js
import { __assign as __assign_1, __spreadArray } from "tslib";
import { isfunc, initializeApi, mergeApis } from 'piral-base';
import { withApi } from '../state';
import { ExtensionSlot } from '../components';
import { createDataOptions, getDataExpiration, renderInDom } from '../utils';
export function createCoreApi(context) {
    return function (api, target) {
        var pilet = target.name;
        return {
            getData: function (name) {
                return context.readDataValue(name);
            },
            setData: function (name, value, options) {
                var _a = createDataOptions(options), _b = _a.target, target = _b === void 0 ? 'memory' : _b, expires = _a.expires;
                var expiration = getDataExpiration(expires);
                return context.tryWriteDataItem(name, value, pilet, target, expiration);
            },
            registerPage: function (route, arg, meta) {
                context.registerPage(route, {
                    pilet: pilet,
                    meta: meta,
                    component: withApi(context, arg, api, 'page'),
                });
                return function () { return api.unregisterPage(route); };
            },
            unregisterPage: function (route) {
                context.unregisterPage(route);
            },
            registerExtension: function (name, arg, defaults) {
                context.registerExtension(name, {
                    pilet: pilet,
                    component: withApi(context, arg, api, 'extension'),
                    reference: arg,
                    defaults: defaults,
                });
                return function () { return api.unregisterExtension(name, arg); };
            },
            unregisterExtension: function (name, arg) {
                context.unregisterExtension(name, arg);
            },
            renderHtmlExtension: function (element, props) {
                var id = renderInDom(context, element, ExtensionSlot, props);
                return function () { return context.destroyPortal(id); };
            },
            Extension: ExtensionSlot,
        };
    };
}
export function createExtenders(context, apis) {
    var creators = __spreadArray([createCoreApi], apis.filter(isfunc));
    return creators.map(function (c) {
        var ctx = c(context);
        if (isfunc(ctx)) {
            return ctx;
        }
        else {
            return function () { return (__assign_1({}, ctx)); };
        }
    });
}
export function defaultApiFactory(context, apis) {
    var extenders = createExtenders(context, apis);
    return function (target) {
        var api = initializeApi(target, context);
        context.apis[target.name] = api;
        return mergeApis(api, extenders, target);
    };
}
