// piral-core/esm/modules/dependencies.js
import { __assign } from "tslib";
import { isfunc } from 'piral-base';
var sharedDependencies = {};
if (process.env.SHARED_DEPENDENCIES) {
    var fillDependencies = require('../../dependencies.codegen');
    if (isfunc(fillDependencies)) {
        fillDependencies(sharedDependencies);
    }
}
/**
 * The global dependencies, which represent the dependencies
 * already used by piral-core itself.
 */
export var globalDependencies = {
    react: require('react'),
    'react-dom': require('react-dom'),
    'react-router': require('react-router'),
    'react-router-dom': require('react-router-dom'),
    history: require('history'),
    tslib: require('tslib'),
    'path-to-regexp': require('path-to-regexp'),
    '@libre/atom': require('@libre/atom'),
    '@dbeining/react-atom': require('@dbeining/react-atom'),
};
/**
 * Gets the local dependencies for the app shell, which
 * are the global dependencies and the implicitly shared
 * dependencies.
 */
export function getLocalDependencies() {
    return __assign(__assign({}, globalDependencies), sharedDependencies);
}
/**
 * The default pilet requester, which just returns an empty array.
 * Use your own code to connect to the feed service.
 *
 * @example
```js
return fetch('https://feed.piral.cloud/api/v1/pilet/sample')
  .then(res => res.json())
  .then(res => res.items);
```
 */
export function defaultModuleRequester() {
    return Promise.resolve([]);
}
