// piral-core/esm/actions/data.js
import { __assign } from "tslib";
import { updateKey } from '../utils';
export function resetData(ctx) {
    ctx.dispatch(function (state) { return (__assign(__assign({}, state), { data: {} })); });
}
export function readDataItem(ctx, key) {
    return ctx.readState(function (state) { return state.data[key]; });
}
export function readDataValue(ctx, key) {
    var item = readDataItem(ctx, key);
    return item && item.value;
}
export function writeDataItem(ctx, key, value, owner, target, expires) {
    var isNull = !value && typeof value === 'object';
    var data = isNull
        ? value
        : {
            value: value,
            owner: owner,
            target: target,
            expires: expires,
        };
    ctx.dispatch(function (state) { return (__assign(__assign({}, state), { data: updateKey(state.data, key, data) })); });
    ctx.emit('store-data', {
        name: key,
        target: target,
        value: value,
        owner: owner,
        expires: expires,
    });
}
export function tryWriteDataItem(ctx, key, value, owner, target, expires) {
    var item = readDataItem(ctx, key);
    if (item && item.owner !== owner) {
        console.error("Invalid data write to '" + key + "'. This item currently belongs to '" + item.owner + "' (write attempted from '" + owner + "'). The action has been ignored.");
        return false;
    }
    writeDataItem(ctx, key, value, owner, target, expires);
    return true;
}
