// piral-core/esm/components/components.js
import { __assign } from "tslib";
import * as React from 'react';
import { useGlobalState } from '../hooks';
export function getPiralComponent(name) {
    return function (props) {
        var Component = useGlobalState(function (s) { return s.components[name]; });
        return Component ? React.createElement(Component, __assign({}, props)) : null;
    };
}
/**
 * Gets the currently registered ErrorInfo component.
 * By default the DefaultErrorInfo component is used.
 */
export var PiralError = getPiralComponent('ErrorInfo');
/**
 * Gets the currently registered LoadingIndicator component.
 * By default only Loading is rendered.
 */
export var PiralLoadingIndicator = getPiralComponent('LoadingIndicator');
/**
 * Gets the currently registered Router component.
 * By default the BrowserRouter is used.
 */
export var PiralRouter = getPiralComponent('Router');
/**
 * Gets the currently registered Layout component.
 * By default the children are rendered.
 */
export var PiralLayout = getPiralComponent('Layout');
/**
 * Gets the currently registered Debug component.
 * By default nothing is used.
 */
export var PiralDebug = getPiralComponent('Debug');
