// piral-base/lib/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createListener = void 0;
function nameOf(type) {
    return "piral-" + type;
}
/**
 * Creates a new Piral app shell event emitter.
 * Uses a custom event dispatcher with a state for usage control.
 * @param state The optional state object to identify the instance.
 * @returns The event emitter.
 */
function createListener(state) {
    if (state === void 0) { state = {}; }
    var eventListeners = [];
    return {
        on: function (type, callback) {
            var listener = function (_a) {
                var detail = _a.detail;
                return detail && detail.state === state && callback(detail.arg);
            };
            document.body.addEventListener(nameOf(type), listener);
            eventListeners.push([callback, listener]);
            return this;
        },
        off: function (type, callback) {
            var listener = eventListeners.filter(function (m) { return m[0] === callback; })[0];
            if (listener) {
                document.body.removeEventListener(nameOf(type), listener[1]);
                eventListeners.splice(eventListeners.indexOf(listener), 1);
            }
            return this;
        },
        emit: function (type, arg) {
            var ce = document.createEvent('CustomEvent');
            ce.initCustomEvent(nameOf(type), false, false, {
                arg: arg,
                state: state,
            });
            document.body.dispatchEvent(ce);
            return this;
        },
    };
}
exports.createListener = createListener;
