// piral-notifications/esm/create.js
import { __assign } from "tslib";
import * as actions from './actions';
import { isValidElement } from 'react';
import { withApi, defaultRender } from 'piral-core';
import { DefaultHost, DefaultToast } from './default';
function isElement(element) {
    return isValidElement(element);
}
function toComponent(component) {
    if (typeof component === 'string') {
        var text_1 = component;
        return function () { return defaultRender(text_1); };
    }
    else if (isValidElement(component)) {
        var element_1 = component;
        return function () { return element_1; };
    }
    return component;
}
function createNotification(context, id, content, defaultOptions, customOptions) {
    if (customOptions === void 0) { customOptions = {}; }
    var options = __assign(__assign({}, defaultOptions), customOptions);
    var notification = {
        id: id,
        component: toComponent(content),
        options: options,
        close: function () {
            setTimeout(function () { return context.closeNotification(notification); }, 0);
        },
    };
    if (typeof options.autoClose === 'number' && options.autoClose > 0) {
        setTimeout(notification.close, options.autoClose);
    }
    return notification;
}
function getNotifications(context, messages, defaultOptions) {
    var notifications = [];
    var i = 0;
    for (var _i = 0, messages_1 = messages; _i < messages_1.length; _i++) {
        var _a = messages_1[_i], content = _a.content, options = _a.options;
        notifications.push(createNotification(context, "global-" + i++, content, defaultOptions, options));
    }
    return notifications;
}
/**
 * Creates new Pilet API extensions for showing notifications.
 */
export function createNotificationsApi(config) {
    if (config === void 0) { config = {}; }
    var _a = config.defaultOptions, defaultOptions = _a === void 0 ? {} : _a, _b = config.selectId, selectId = _b === void 0 ? function () { return "" + ~~(Math.random() * 10000); } : _b, _c = config.messages, messages = _c === void 0 ? [] : _c;
    return function (context) {
        context.defineActions(actions);
        context.dispatch(function (state) { return (__assign(__assign({}, state), { components: __assign({ NotificationsHost: DefaultHost, NotificationsToast: DefaultToast }, state.components), notifications: getNotifications(context, messages, defaultOptions) })); });
        return function (api) { return ({
            showNotification: function (content, customOptions) {
                var Component = typeof content === 'string'
                    ? content
                    : isElement(content)
                        ? content
                        : withApi(context, content, api, 'extension');
                var notification = createNotification(context, selectId(), Component, defaultOptions, customOptions);
                context.openNotification(notification);
                return notification.close;
            },
        }); };
    };
}
