// piral-core/lib/modules/dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultModuleRequester = exports.getLocalDependencies = exports.globalDependencies = void 0;
var tslib_1 = require("tslib");
var piral_base_1 = require("piral-base");
var sharedDependencies = {};
if (process.env.SHARED_DEPENDENCIES) {
    var fillDependencies = require('../../dependencies.codegen');
    if (piral_base_1.isfunc(fillDependencies)) {
        fillDependencies(sharedDependencies);
    }
}
/**
 * The global dependencies, which represent the dependencies
 * already used by piral-core itself.
 */
exports.globalDependencies = {
    react: require('react'),
    'react-dom': require('react-dom'),
    'react-router': require('react-router'),
    'react-router-dom': require('react-router-dom'),
    history: require('history'),
    tslib: require('tslib'),
    'path-to-regexp': require('path-to-regexp'),
    '@libre/atom': require('@libre/atom'),
    '@dbeining/react-atom': require('@dbeining/react-atom'),
};
/**
 * Gets the local dependencies for the app shell, which
 * are the global dependencies and the implicitly shared
 * dependencies.
 */
function getLocalDependencies() {
    return tslib_1.__assign(tslib_1.__assign({}, exports.globalDependencies), sharedDependencies);
}
exports.getLocalDependencies = getLocalDependencies;
/**
 * The default pilet requester, which just returns an empty array.
 * Use your own code to connect to the feed service.
 *
 * @example
```js
return fetch('https://feed.piral.cloud/api/v1/pilet/sample')
  .then(res => res.json())
  .then(res => res.items);
```
 */
function defaultModuleRequester() {
    return Promise.resolve([]);
}
exports.defaultModuleRequester = defaultModuleRequester;
